cpp/src/arrow/compute/light_array.h

```diff
@@ -40,9 +40,35 @@ namespace compute {
 /// allows us to take advantage of these resources without coupling the logic with
 /// the execution engine.
 struct LightContext {
+  static constexpr int kLogMinibatchSizeMax = 10;
+  static constexpr int kMinibatchSizeMax = 1 << kLogMinibatchSizeMax;
+
   bool has_avx2() const { return (hardware_flags & arrow::internal::CpuInfo::AVX2) > 0; }
   int64_t hardware_flags;
   util::TempVectorStack* stack;
+  MemoryPool* memory_pool;
+
+  // A deleter for a light context that owns its TempVectorStack
+  struct OwningLightContextDeleter {
+    void operator()(LightContext* ctx) const {
+      delete ctx->stack;
+      delete ctx;
+    };
+  };
+
+  // Creates a new light context with an owned temp vector stack.  Only to be
+  // used at the highest level (e.g. in ExecPlan, unit tests or benchmarks)
+  // Nodes should get their light context from the plan
```

Comment:
I think generally people will rely on ExecContext (or QueryContext once that merges in). LightContext is more designed to be created on the stack and passed around to lower-level code. The biggest use-case is to not have to continuously look up the thread-local TempVectorStack and instead look it up once and cache it here.

Reply:
We have too many context objects at the moment :laughing: .  `LightContext` was my poor attempt at https://xkcd.com/927/

However, since it holds a non-owning pointer to `TempVectorStack` it can't just be easily created for unit tests.  So this was just a quick-hack for unit tests.  The unique_ptr with a custom deleter is just a way of doing...

```
struct LightContextThatAlsoOwnsTempVectorStack {
  LightContext ctx;
  TempVectorStack stack;
}
```

...without creating a new class since I was afraid of creating "yet another context variant".  However, I think I might as well do that explicitly and create a new class.  It will be less confusing to new readers.

Reply:
However, I completely agree that once `QueryContext` merges in then this entire construct can go away.  I can't get away with `ExecContext` because it doesn't have `TempVectorStack`.

Reply:
Right but then something like this...

```
TEST(SomeTest, ThatNeedsLightContext) {
  auto ctx = LightContext::Make();
  DoThingWithCtx(*ctx);
}
```

...becomes...

```
TEST(SomeTest, ThatNeedsLightContext) {
  util::TempVectorStack stack;
  stack.Init(kWhatWasThatParameterAgain);
  LightContext ctx(&stack);
  DoThingWithCtx(ctx);
}
```

Whatever our "master context" ends up being I would very much like to have a one-liner that creates one and all I have to do is keep track of the thing that comes back from that one-liner.  Otherwise I have to remember the recipe every time I write a unit test.

Reply:
Ah makes sense, if we're getting rid of it anyway after QueryContext then it's probably fine